File: src/main/scala/com/exasol/cloudetl/bucket/SecureBucket.scala
package com.exasol.cloudetl.bucket

import com.exasol.errorreporting.ExaError

import com.typesafe.scalalogging.LazyLogging

/**
 * A trait that provides methods to access [[Bucket]]s using secure
 * access credentials.
 */
trait SecureBucket extends LazyLogging { self: Bucket =>

  /**
   * Return the list of property key names that are used as secure
   * access credentials.
   *
   * For example, {@code AWS_SECRET_KEY} when accessing an S3 bucket.
   */
  def getSecureProperties(): Seq[String]

  /**
   * Validates that the named connection object or access credentials
   * are available.
   */
  protected[this] final def validateConnectionProperties(): Unit = {
    if (hasSecureProperties()) {
      throw new BucketValidationException(
        ExaError
          .messageBuilder("E-CSE-5")
          .message("Using credentials as parameters is forbidded.")
          .mitigation("Please use an Exasol named connection object via CONNECTION_NAME property.")
          .toString()
      )
    }
    if (!properties.hasNamedConnection()) {
      throw new BucketValidationException(
        ExaError
          .messageBuilder("E-CSE-6")
          .message("No CONNECTION_NAME property is defined.")
          .mitigation("Please use connection object to provide access credentials.")
          .toString()
      )
    }
  }

  private[this] def hasSecureProperties(): Boolean =
    getSecureProperties.exists(properties.containsKey(_))

}

File: src/main/java/com/exasol/cloudetl/bucket/SecureBucket.java
package com.exasol.cloudetl.bucket;

import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.storage.StorageProperties;
import com.exasol.errorreporting.ExaError;

/** Mixin interface for buckets that require secure credentials. */
public interface SecureBucket {
    /** @return secure property keys */
    scala.collection.immutable.Seq<String> getSecureProperties();

    /** @return storage properties */
    StorageProperties properties();

    /** Validate connection credential properties. */
    default void validateConnectionProperties() {
        if (hasSecureProperties()) {
            throw new BucketValidationException(ExaError.messageBuilder("E-CSE-5")
                    .message("Using credentials as parameters is forbidded.")
                    .mitigation("Please use an Exasol named connection object via CONNECTION_NAME property.").toString());
        }
        if (!properties().hasNamedConnection()) {
            throw new BucketValidationException(ExaError.messageBuilder("E-CSE-6")
                    .message("No CONNECTION_NAME property is defined.")
                    .mitigation("Please use connection object to provide access credentials.").toString());
        }
    }

    private boolean hasSecureProperties() {
        for (final String key : ScalaConverters.asJavaList(getSecureProperties())) {
            if (properties().containsKey(key)) {
                return true;
            }
        }
        return false;
    }
}
